habitat/tasks/nav/multi_object_nav_task.py

```diff
@@ -0,0 +1,551 @@
+        episode: MultiObjectGoalNavEpisode,
+        task: MultiObjectNavigationTask,
+        **kwargs: Any,
+    ):
+        current_position = self._sim.get_agent_state().position.tolist()
+        distance_to_current_goal = self._sim.geodesic_distance(
+            current_position, episode.goals[task.current_goal_index].position
+        )
+
+        self._metric = distance_to_current_goal
+
+
+@registry.register_measure
+class CurrentGoalSuccess(Measure):
+    r"""Whether or not the agent succeeded in finding it's
+    current goal. This measure depends on DistanceToCurrentObjectGoal measure.
+    """
+
+    cls_uuid: str = "current_goal_success"
+
+    def __init__(
+        self, *args: Any, sim: Simulator, config: Config, **kwargs: Any
+    ):
+        self._sim = sim
+        self._config = config
+
+        super().__init__()
+
+    def _get_uuid(self, *args: Any, **kwargs: Any):
+        return self.cls_uuid
+
+    def reset_metric(self, *args: Any, episode, task, **kwargs: Any):
+        task.measurements.check_measure_dependencies(
+            self.uuid, [DistanceToCurrentObjectGoal.cls_uuid]
+        )
+        self.update_metric(*args, episode=episode, task=task, **kwargs)
+
+    def update_metric(
+        self,
+        *args: Any,
+        episode,
+        task: MultiObjectNavigationTask,
+        **kwargs: Any,
+    ):
+        distance_to_current_goal = task.measurements.measures[
+            DistanceToCurrentObjectGoal.cls_uuid
+        ].get_metric()
+
+        if (
+            task.is_found_called
+            and distance_to_current_goal <= self._config.SUCCESS_DISTANCE
+        ):
+            self._metric = 1
+            task.current_goal_index += 1
+        else:
+            self._metric = 0
+
+
+@registry.register_measure
+class Progress(Measure):
+    r"""What fraction of goals are found. This measure depends on
+    CurrentGoalSuccess measure.
+    """
+
+    cls_uuid: str = "progress"
+
+    def __init__(
+        self, *args: Any, sim: Simulator, config: Config, **kwargs: Any
+    ):
+        self._sim = sim
+        self._config = config
+
+        super().__init__()
+
+    def _get_uuid(self, *args: Any, **kwargs: Any):
+        return self.cls_uuid
+
+    def reset_metric(self, *args: Any, episode, task, **kwargs: Any):
+        task.measurements.check_measure_dependencies(
+            self.uuid, [CurrentGoalSuccess.cls_uuid]
+        )
+        self._metric = 0
+        self.update_metric(*args, episode=episode, task=task, **kwargs)
+
+    def update_metric(
+        self,
+        *args: Any,
+        episode: MultiObjectGoalNavEpisode,
+        task: MultiObjectNavigationTask,
+        **kwargs: Any,
+    ):
+        ep_current_goal_success = task.measurements.measures[
+            CurrentGoalSuccess.cls_uuid
+        ].get_metric()
+
+        if ep_current_goal_success:
+            self._metric += 1 / len(episode.goals)
+
+
+@registry.register_measure
+class MultiONSuccess(Measure):
```

Comment:
In general, I don't think I like this design of task specific measures for metrics that are also well defined in other tasks.  IMO MultiON should just use Success and then that class should dispatch to different implementations based on the task.  In the same way that the topdown map and DistanceToGoal measures currently work for objectnav and pointnav.

Reply:
I am a bit hesitant of the idea of merging MultiON specific MultiONSuccess (and DistanceToCurrentObjectGoal) measures to the previously defined Success (and DistanceToGoal) measures because of the following reasons -
1. This is not scalable. If and when other works will be merged to habitat, all task-specific metrics can't be eventually merged into the existing metric definition.
2. There will be typing conflicts as other works may refer to different episode[[1](https://github.com/facebookresearch/habitat-lab/blob/master/habitat/tasks/nav/nav.py#L962)] and task[[2](https://github.com/facebookresearch/habitat-lab/blob/master/habitat/tasks/nav/nav.py#L519)] class definition. In our case, episode is `MultiObjectGoalNavEpisode` and task is `MultiObjectNavigationTask`.
3. Much of the logic in Success and DistancetoCurrentGoal will pertain to MultiON. This brings an unnecessary overhead for someone trying to work with PointNav/ObjectNav tasks in Habitat.  

This said, I am happy to follow whatever you recommend.

Reply:
Concerns 1 and 2 are largely implementation questions that can both be largely resolved by constructing different implementations of a given measure based on the task.  A stack of `if`'s isn't scalable or nice, but we can use other patterns to get around that without framing out that work to the user.

I do see 3, another layer of indirection will make the implementation more confusing.  So it's a question of whether a more confusing implementation is worth a cleaner API.  @mathfac @Skylion007 do either of you have thoughts here?